### PR TITLE
Precompute and store attribute and executor id together in a single h…

### PR DESCRIPTION
**Pull request by baldersheim**
…ashmap.

the executor id so complex fields use the same executor.
@toregge and @geirst PR

### CODE REVIEW

File: searchcore/src/vespa/searchcore/proton/attribute/attribute_writer.cpp [hunk 1]
```diff
@@ -474,9 +474,15 @@ AttributeWriter::AttributeWriter(const proton::IAttributeManager::SP &mgr)
...
       _dataType(nullptr),
-      _hasStructFieldAttribute(false)
+      _hasStructFieldAttribute(false),
+      _attrMap()
 {
     setupWriteContexts();
+    for (auto attr : getWritableAttributes()) {
+        vespalib::stringref name = attr->getName();
+        vespalib::stringref prefix = name.substr(0, name.find('.'));
+        _attrMap[attr->getName()] = AttrWithId(attr, _attributeFieldWriter.getExecutorId(prefix));
```
Comment (toregge) [inline searchcore/src/vespa/searchcore/proton/attribute/attribute_writer.cpp]: This doesn't match how executor id is set in FieldContext constructor.

**Review by toregge — CHANGES_REQUESTED** [review]
Please ensure consistent mapping to executor id.

**Comment by baldersheim** [comment]
Unified selection of executorId.

**Comment by toregge** [comment]
Doesn't match executor id used in AttributeWriter::heartBeat(), AttributeWriter::onReplayDone(),
AttributeWriter::compactLidSpace(), FlushableAttribute::initFlush(), ExclusiveAttributeReadAccessor::takeGuard(), AttributeManager::asyncForEachAttribute(), 
FilterAttributeManager::asyncForEachAttribute(), GidToLidChangeListener::GidToLidChangeListener(),
and allocShrinker() (not necessarily a complete list).

**Comment by baldersheim** [comment]
Now, I think I have caught all of them ..

**Review by toregge — APPROVED** [review]
LGTM

File: searchcore/src/vespa/searchcore/proton/attribute/attribute_writer.h [hunk 2]
```diff
@@ -58,11 +58,15 @@ class AttributeWriter : public IAttributeWriter
...
 private:
+    using AttrWithId = std::pair<search::AttributeVector *, ExecutorId>;
+    using AttrMap = vespalib::hash_map<vespalib::string, AttrWithId>;
     std::vector<WriteContext> _writeContexts;
     const DataType           *_dataType;
     bool                      _hasStructFieldAttribute;
+    AttrMap                   _attrMap;
 
     void setupWriteContexts();
+    void setupAttriuteMapping();
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/attribute/attribute_writer.h]: Typo: Attriute -> Attribute

**Review by geirst — COMMENTED** [review]
@baldersheim FYI

**Pull request merged**